From: Samuel Thibault <samuel.thibault@ens-lyon.org>
Subject: [PATCH] Add fork hooks for pthread_atfork

2013-09-21  Samuel Thibault  <samuel.thibault@ens-lyon.org>

pthread_atfork needs application callbacks to be called outside any locking.

* sysdeps/mach/hurd/fork.c (_hurd_atfork_prepare_hook, _hurd_atfork_child_hook,
_hurd_atfork_parent_hook): New hooks.
(__fork): Call _hurd_atfork_prepare_hook hooks before all locking, call
_hurd_atfork_parent_hook or _hurd_atfork_child_hook after all unlocking.

Signed-off-by: Samuel Thibault <samuel.thibault@ens-lyon.org>

---
 sysdeps/mach/hurd/fork.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/sysdeps/mach/hurd/fork.c b/sysdeps/mach/hurd/fork.c
index ab11bab..3a6803e 100644
--- a/sysdeps/mach/hurd/fork.c
+++ b/sysdeps/mach/hurd/fork.c
@@ -34,6 +34,11 @@
 symbol_set_declare (_hurd_fork_locks)
 
 
+/* Application callbacks registered through pthread_atfork.  */
+DEFINE_HOOK (_hurd_atfork_prepare_hook, (void));
+DEFINE_HOOK (_hurd_atfork_child_hook, (void));
+DEFINE_HOOK (_hurd_atfork_parent_hook, (void));
+
 /* Things that want to be called before we fork, to prepare the parent for
    task_create, when the new child task will inherit our address space.  */
 DEFINE_HOOK (_hurd_fork_prepare_hook, (void));
@@ -62,6 +67,8 @@ __fork (void)
   error_t err;
   struct hurd_sigstate *volatile ss;
 
+  RUN_HOOK (_hurd_atfork_prepare_hook, ());
+
   ss = _hurd_self_sigstate ();
   __spin_lock (&ss->critical_section_lock);
 
@@ -695,6 +702,14 @@ __fork (void)
 
   _hurd_critical_section_unlock (ss);
 
+  if (!err)
+    {
+      if (pid)
+	RUN_HOOK (_hurd_atfork_parent_hook, ());
+      else
+	RUN_HOOK (_hurd_atfork_child_hook, ());
+    }
+
   return err ? __hurd_fail (err) : pid;
 }
 libc_hidden_def (__fork)
-- 
tg: (9a079e2..) t/pthread-atfork (depends on: baseline)
